[CL-2156] Only send ProjectPhaseStared mail to Users who participated in Project

jinjagit opened the pull request:
This PR changes the query central to `def self.make_notifications_on` in `back/app/models/notifications/project_phase_started.rb` so that we only notify `Users` who have participated in the relevant `Project` that a `Phase` has started, and who also can still participate in the project.  

Since [campaigns/project_phase_started.rb](https://github.com/CitizenLabDotCo/citizenlab/blob/581391b899c32876314e45f37d9fd654cafc27a7/back/engines/free/email_campaigns/app/models/email_campaigns/campaigns/project_phase_started.rb#L49) is `ActivityTriggerable`, this will result in the email campaign for this event being sent to each `User` who received the above-mentioned `Notification`.

Please [see ticket](https://citizenlab.atlassian.net/browse/CL-2156) for discussion of how the audience of `ProjectPhaseUpcoming` notifications/mails complements the new audience (due to this PR) of `ProjectPhaseStarted` notifications/mails.

jinjagit committed:
[CL-2156] Basic fix & test

Commit:
Translations updated by CI (extract-intl)

jinjagit committed:
[CL-2156] Temporary fix of tests - likely to change in light of Product responses to my questions

jinjagit committed:
Merge branch 'CL-2156-only-send-new-phase-email-to-users-who-participated' of github.com:CitizenLabDotCo/citizenlab into CL-2156-only-send-new-phase-email-to-users-who-participated

jinjagit committed:
Merge branch 'master' into CL-2156-only-send-new-phase-email-to-users-who-participated

jinjagit committed:
[CL-2156] Temprarily comment out unused vars in test

jinjagit committed:
[CL-2156] Remove consideration of moderation role from model and tests

jinjagit committed:
[CL-2156] Remove project_phase_started_spec, as no longer useful

jinjagit committed:
[CL-2156] Update changelog

jinjagit committed:
Merge branch 'master' into CL-2156-only-send-new-phase-email-to-users-who-participated

jinjagit committed:
[CL-2156] Use let statements in describe block

jinjagit committed:
Merge branch 'CL-2156-only-send-new-phase-email-to-users-who-participated' of github.com:CitizenLabDotCo/citizenlab into CL-2156-only-send-new-phase-email-to-users-who-participated

jinjagit commented on back/spec/models/notification_spec.rb:
```diff
@@ -87,49 +87,21 @@
+      let(:participant) { create(:user) }
+      let(:non_participant) { create(:user) }
+
+      let(:idea) { create(:idea, project_id: project.id, author_id: participant.id) }
```
Maybe I should take this opportunity to use let statements in the existing tests in this spec too? Thoughts?

alexander-cit replied:
I think `let` is more useful for sharing variables between different specs. For one spec, local variables are fine.

alexander-cit commented on back/spec/models/notification_spec.rb:
```diff
@@ -87,49 +87,21 @@
-        phase = create(:active_phase)
-        project = phase.project
-        project.visible_to = 'groups'
-        project.groups << create(:group)
```
So, you're removing all the tests related to groups, but the group logic is still in `ProjectPhaseStarted.make_notifications_on`. Why do you think we don't need to test it anymore?

jinjagit replied:
Because the latter two tests are about which moderator gets the notification. Now no moderator does (not because they are a moderator), only users who participated get the notification.

The first test is a test of groups, but implicitly this is tested by participation, since you should not be able to participate unless you have permission (in a group specified for project, for example). Which makes me wonder if we even need `participation_possible_for_context?` in the `ProjectPhaseStarted.make_notifications_on` method :thinking:

jinjagit replied:
I guess we do need it, since a user could participate in a project before permissions (groups, etc.) are added/changed and therefore not be able to participate by the time the notification is processed.

alexander-cit replied:
> I guess we do need it,

Yes, I agree.

jinjagit replied:
Thanks, I hadn't thought of that case :+1:

alexander-cit replied:
Ah I've figured it out. You don't need to use groups in your test because you don't set `project.visible_to = 'groups'`. Maybe indeed it's better not to test groups here as it's another class's responsibility. 

But at the same time, `visible_to: 'groups'` is not tested here `back/spec/policies/project_policy_spec.rb`. So, maybe this spec is the only place where it was checked. 

I mean it's checked by this line (the variable name is wrong and confusing)

```ruby
_other_admin = create(:user, email: 'user2@example.com')
```

`user` is created with groups, and so it's included in `expect`, contrary to `_other_admin`.

alexander-cit commented on back/spec/models/notifications/project_phase_started_spec.rb:
```diff
@@ -1,33 +0,0 @@
-      activity = create :activity, item: phase, action: 'started'
-      create_list :user, 2
-      allow_any_instance_of(ParticipationContextService).to(
-        receive(:participation_possible_for_context?).and_return(true, false)
```
Now we don't test the influence of `participation_possible_for_context?` on `make_notifications_on`, but again the check is still there. Why do you think we don't need to test it?

jinjagit replied:
Hmm. Good point. I'll replace this test.

alexander-cit replied:
If it does its job, you can just leave it here.
I would also move the `ProjectPhaseStarted` specific tests from `back/spec/models/notification_spec.rb` to this file (if you're going to refactor it anyway). Not sure why we test different classes in one file.

alexander-cit commented on back/spec/models/notification_spec.rb:
```diff
@@ -87,49 +87,21 @@
+      it 'only notifies users who participated in project' do
+        create(:activity, item: idea, user_id: participant.id, action: 'published')
+        notifications = Notifications::ProjectPhaseStarted.make_notifications_on activity
+        notification_ids = notifications.pluck(:recipient_id)
```
What do you think about naming it `recipient_ids`?

alexander-cit commented on back/app/models/notifications/project_phase_started.rb:
```diff
@@ -65,7 +65,7 @@ def self.make_notifications_on(activity)
       if phase.project
-        user_scope = User.where.not(id: UserRoleService.new.moderators_for(phase))
+        user_scope = ParticipantsService.new.projects_participants(Project.where(id: phase.project_id))
         ActiveRecord::Base.transaction do
```
This transaction (and the comment) is really confusing. We don't create anything here since this commit https://github.com/CitizenLabDotCo/citizenlab/commit/39508eb, so we don't need the transaction.

If you have time/eagerness, could you remove it?

jinjagit replied:
It was only added 2 weeks ago, by Sebi, so I suspect it is needed.

alexander-cit replied:
I believe it was added three years ago https://github.com/CitizenLabDotCo/citizenlab/commit/f47e37c

jinjagit replied:
Sorry, I meant 'touched by' Sebi 2 weeks ago.

jinjagit replied:
I can remove it, if not needed, for sure.

jinjagit committed:
[CL-2156] Add test of can participate and move tests to project_phase_started_spec

jinjagit committed:
[CL-2156] Simplify test

jinjagit committed:
[CL-2156] remove unnecessary user attributes in test

jinjagit committed:
[CL-2156] Remove redundant ActiveRecord transaction

jinjagit commented:
@alexander-cit Thanks for your review. Your suggestions were really good. I have incorporated all changes you suggested.

1. I removed the redundant `ActiveRecord` transaction.
2. I moved all related tests back into the relevant model spec: `project_phase_started_spec.rb`
3. I replaced a test (modified to use some common variables) that only Users who _can_ participate are notified. So, now we test only project participants who _also can still participate_ get notified.

jinjagit committed:
Merge branch 'master' into CL-2156-only-send-new-phase-email-to-users-who-participated

jinjagit commented on back/spec/models/notification_spec.rb:
```diff
@@ -86,81 +86,6 @@
-      end
-    end
-
-    describe 'project_phase_upcoming' do
-      let(:project) { create(:project_with_current_phase) }
-      let(:other_project) { create(:project_with_current_phase) }
-
-      let(:project_folder) { create(:project_folder, projects: [project]) }
-      let(:other_project_folder) { create(:project_folder, projects: [other_project]) }
-
-      let!(:regular_user) { create(:user) }
-      let!(:moderator_of_project) { create(:project_moderator, projects: [project]) }
-      let!(:moderator_of_other_project) { create(:project_moderator, projects: [other_project]) }
-      let!(:moderator_of_project_folder) { create(:project_folder_moderator, project_folders: [project_folder]) }
-      let!(:moderator_of_other_project_folder) { create(:project_folder_moderator, project_folders: [other_project_folder]) }
-      let!(:admin) { create(:admin) }
-
-      it 'notifies all admins and only relevant moderators' do
-        activity = create(:activity, item: project.phases[2], action: 'upcoming')
-        notifications = Notifications::ProjectPhaseUpcoming.make_notifications_on activity
-        notification_ids = notifications.pluck(:recipient_id)
-
-        expect(notification_ids).to include moderator_of_project.id
-        expect(notification_ids).to include moderator_of_project_folder.id
-        expect(notification_ids).to include admin.id
-        expect(notification_ids).not_to include regular_user.id
-        expect(notification_ids).not_to include moderator_of_other_project.id
-        expect(notification_ids).not_to include moderator_of_other_project_folder.id
-      end
-    end
```
Removed from this file, as I forgot to do this in #3366 (when I moved the test to `project_phase_upcoming_spec.rb`, but did not remove the original here)

jinjagit committed:
Merge branch 'master' into CL-2156-only-send-new-phase-email-to-users-who-participated

jinjagit committed:
Merge branch 'CL-2156-only-send-new-phase-email-to-users-who-participated' of github.com:CitizenLabDotCo/citizenlab into CL-2156-only-send-new-phase-email-to-users-who-participated
